context/app/static/js/components/entity-search/sidebar/DatasetConfigureSearchTable/hooks.js

```diff
@@ -0,0 +1,96 @@
+
+function selectedItemsReducer(state, { type, payload }) {
+  const tempState = state;
+  switch (type) {
+    case 'selectItem':
+      return { ...state, [payload]: undefined }; // using an object as a set
+    case 'deselectItem':
+      delete tempState[payload];
+      return { ...tempState };
+    case 'setSelectedItems':
+      return payload;
+    default:
+      return state;
+  }
+}
+
+function useGroupedMetadataFieldConfigs() {
+  const metadataFieldConfigs = useMetadataFieldConfigs();
+  return Object.entries(metadataFieldConfigs).reduce((acc, [metadataFieldName, metadataFieldConfig]) => {
+    return produce(acc, (draft) => {
+      // eslint-disable-next-line no-param-reassign
+      draft[metadataFieldConfig.facetGroup] = {
+        ...draft[metadataFieldConfig.facetGroup],
+        [metadataFieldName]: metadataFieldConfig,
+      };
+      return draft;
+    });
+  }, {});
+}
+
+function getSelectedGroupsFieldConfigs(groups, selectedGroups) {
+  return Object.keys(selectedGroups).reduce((acc, selectedGroup) => {
+    return { ...acc, ...groups[selectedGroup] };
+  }, {});
+}
+
+function useFieldGroups() {
+  const { initialFields, initialFacets } = useStore();
+  const groupedMetadataFieldConfigs = useGroupedMetadataFieldConfigs();
+  const groups = { General: { ...initialFields, ...initialFacets }, ...groupedMetadataFieldConfigs };
+  const { selectedItems: selectedGroups, handleToggleItem: handleToggleGroup } = useSelectedItems(
+    selectedItemsReducer,
+    {},
+  );
+
+  const selectedGroupFieldConfigs = getSelectedGroupsFieldConfigs(groups, selectedGroups);
+  return { groups, selectedGroups, handleToggleGroup, selectedGroupFieldConfigs };
+}
+
+function buildFieldConfigs(fieldNames) {
+  return fieldNames.reduce(
+    (acc, fieldName) => ({
+      ...acc,
+      ...createDatasetFacet({ fieldName, entityType: 'dataset' }),
+    }),
+    {},
+  );
+}
+
+function getDataTypesFields(selectedDataTypes) {
+  return Object.keys(selectedDataTypes).reduce((acc, selectedDataType) => {
+    const dataTypeFieldConfigs = buildFieldConfigs(dataTypesToFieldsMap[selectedDataType]);
+    return { ...acc, ...dataTypeFieldConfigs };
+  }, {});
+}
+
+function useSelectedDataTypes() {
+  const { selectedItems: selectedDataTypes, handleToggleItem: handleToggleDataType } = useSelectedItems(
+    selectedItemsReducer,
+    {},
+  );
+
+  const dataTypesFields = getDataTypesFields(selectedDataTypes);
+
+  return { selectedDataTypes, handleToggleDataType, dataTypesFields };
+}
+
+function useDatasetConfigureSearchTable() {
+  const { handleToggleDataType, dataTypesFields } = useSelectedDataTypes();
+  const { groups, handleToggleGroup, selectedGroupFieldConfigs } = useFieldGroups();
+
+  const availableFieldConfigs = { ...selectedGroupFieldConfigs, ...dataTypesFields };
+  return { dataTypesToFieldsMap, handleToggleDataType, groups, handleToggleGroup, availableFieldConfigs };
+}
+
+export { useDatasetConfigureSearchTable };
```

Comment:
I'm not really following this... but this might be a case where it make more sense in context: if I was coming in from the top, I might understand better how each of these is being used.

Reply:
Is the logic of the hooks complicated enough that they should have some unit tests of their own?

Reply:
One react practice that is encourage is encapsulating all of your hooks into a single hook which is something I did less early on, but I am working towards now. I have made some unit tests, and there is a room for more, but where we are under a time pressure, I might revisit once the feature is more whole.